### PR TITLE
[tools] Add coq_flock, a minimal flock replacement for those systems where it is not standard

### LINKED ISSUE #14271: `opam-depext coq.dev` on MacOS should install flock
Since the dune PR, `opam-depext coq.dev` is no longer sufficient to install system dependencies.

### PR DESCRIPTION
**Pull request by ejgallego**
It turns out that flock is a hard to get util in OSX, we thus provide
a basic local implementation as a workaround.

This is a RFC

### CODE REVIEW

**Comment by ejgallego** [comment]
Seems to work (TM) ; folks, what do you think of this approach?

If you think it'd be worth pursuing, I can clean up the code and make this into a mergeable PR.

**Comment by MSoegtropIMC** [comment]
I think this is the most viable solution. I probably would have written it in C right away, but the C + OCaml solution is of course also fine and probably more appropriate in the Coq ecosystem.

I don't see another good solution. Using `shlock` on macOS would require busy waiting (so OK, but not good) and resorting to `-j 1` on macOS is not really a viable option (= very bad). On MacPorts, flock was requested a few times, but it was not added. If we want it in MacPorts, we would have to do it. I am not sure about Homebrew support. It has a flock package but it is something else ("Business messaging and team collaboration app"). Anyway even if Homebrew would have it, dropping MacPorts support cause of this also doesn't make a lot of sense.

**Comment by ppedrot** [comment]
Stupid remark: `clock` is a suprising acronym that reminds me of time measure more than synchronicity barriers, you should probably call it something else or I'll be confused every time I find this in a grep.

**Comment by ejgallego** [comment]
> Stupid remark: `clock` is a suprising acronym that reminds me of time measure more than synchronicity barriers, you should probably call it something else or I'll be confused every time I find this in a grep.

Indeed, I realized about it after the fact, better to rename it to coq_flock , anyways this is a short-term thing we need to maintain as one of the 2 cases should happen soon: dune becomes reentrant, so it is safe to call it in parallel, or we will improve the briding / legacy setup so no locking is needed anymore.

**Comment by ejgallego** [comment]
@MSoegtropIMC this is meant as a band-aid, so indeed this solution is IMHO the most transparent for users.

A big motivation to switch to dune was to make it much easier to use OCaml binaries helper tools [and the compositional nature makes dropping existing stuff easy]

**Comment by JasonGross** [comment]
On my Windows machine:
<details><summary><code>d:/cygwin64/usr/include/sys/file.h</code></summary>

```c
/* This is file FILE.H */
/*
** Copyright (C) 1991 DJ Delorie, 24 Kirsten Ave, Rochester NH 03867-2954
**
** This file is distributed under the terms listed in the document
** "copying.dj", available from DJ Delorie at the address above.
** A copy of "copying.dj" should accompany this file; if not, a copy
** should be available from where this file was obtained.  This file
** may not be distributed without a verbatim copy of "copying.dj".
**
** This file is distributed WITHOUT ANY WARRANTY; without even the implied
** warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
*/
/*
 * 2003-11-27  Nicholas Wourms  <nwourms@netscape.net>:
 *
 *   Include sys/cdefs.h. Add function prototype for flock().
 *   Add some comments from BSD's header for further clarity.
 *   (L_SET, L_CURR, L_INCR, L_XTND): Redefine as the macros
 *   SEEK_SET, SEEK_CUR, SEEK_CUR, & SEEK_END respectively.
 *   (LOCK_SH,LOCK_EX,LOCK_NB,LOCK_UN): New macros for flock().
*/
#ifndef _FILE_H_
#define _FILE_H_

#include <fcntl.h>

/* Whence values for lseek(); renamed by POSIX 1003.1 */
#define L_SET		SEEK_SET
#define L_CURR		SEEK_CUR
#define L_INCR		SEEK_CUR
#define L_XTND		SEEK_END

/* Including <sys/file.h> always defines flock & macros. */
#if __BSD_VISIBLE - 0 == 0

#define LOCK_SH		0x01		/* shared file lock */
#define LOCK_EX		0x02		/* exclusive file lock */
#define LOCK_NB		0x04		/* don't block when locking */
#define LOCK_UN		0x08		/* unlock file */

#ifdef __cplusplus
extern "C"
{
#endif

extern int flock (int, int);

#ifdef __cplusplus
}
#endif

#endif

#endif
```
</details>

<details><summary><code>/usr/include/x86_64-linux-gnu/sys/file.h</code></summary>

```c
/* Copyright (C) 1991-2020 Free Software Foundation, Inc.
   This file is part of the GNU C Library.

   The GNU C Library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   The GNU C Library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with the GNU C Library; if not, see
   <https://www.gnu.org/licenses/>.  */

#ifndef	_SYS_FILE_H
#define	_SYS_FILE_H	1

#include <features.h>

#ifndef	_FCNTL_H
# include <fcntl.h>
#endif

__BEGIN_DECLS


/* Alternate names for values for the WHENCE argument to `lseek'.
   These are the same as SEEK_SET, SEEK_CUR, and SEEK_END, respectively.  */
#ifndef L_SET
# define L_SET	0	/* Seek from beginning of file.  */
# define L_INCR	1	/* Seek from current position.  */
# define L_XTND	2	/* Seek from end of file.  */
#endif


/* Operations for the `flock' call.  */
#define	LOCK_SH	1	/* Shared lock.  */
#define	LOCK_EX	2 	/* Exclusive lock.  */
#define	LOCK_UN	8	/* Unlock.  */

/* Can be OR'd in to one of the above.  */
#define	LOCK_NB	4	/* Don't block when locking.  */


/* Apply or remove an advisory lock, according to OPERATION,
   on the file FD refers to.  */
extern int flock (int __fd, int __operation) __THROW;


__END_DECLS

#endif /* sys/file.h  */
```
</details>

**Comment by JasonGross** [comment]
When I do
```
(echo '#include <sys/file.h>'; echo 'test = (__BSD_VISIBLE,LOCK_SH);') > foo.c; gcc -E foo.c
```
on cygwin, I'm told that `__BSD_VISIBLE` is 1.  On WSL2 it's not defined.

**Comment by JasonGross** [comment]
And `/usr/include/sys/features.h` on cygwin has
```c
#ifdef _DEFAULT_SOURCE
#define __BSD_VISIBLE           1
#else
#define __BSD_VISIBLE           0
#endif
```

**Comment by JasonGross** [comment]
On the other hand, there's
```c
#if __BSD_VISIBLE
/* lock operations for flock(2) */
#define LOCK_SH         0x01            /* shared file lock */
#define LOCK_EX         0x02            /* exclusive file lock */
#define LOCK_NB         0x04            /* don't block when locking */
#define LOCK_UN         0x08            /* unlock file */
#endif
```
in `/usr/include/sys/_default_fcntl.h` which gets included on cygwin, so idk why it's not enabled

**Comment by JasonGross** [comment]
Oh, I'm not reading carefully.  `(echo '#include <sys/file.h>'; echo 'test = LOCK_SH;') > foo.c; gcc -E foo.c` says that `LOCK_SH` is `0x01` on cygwin and `1` on WSL2.  So it should be defined.

**Comment by ejgallego** [comment]
I guess I'll have to fire my windows VM to see what's going on. @JasonGross you can just copy the folder `tools/flock` to some other directoy, add a minimal dune-project and do `dune build @all` 

Does that work?

Minimal dune project file: `(lang dune 2.7)` for example.

**Comment by JasonGross** [comment]
@ejgallego My point is that I don't get the errors you're seeing, so it must be something specific to the version of cygwin you're using / the packages you install.

**Comment by ejgallego** [comment]
> @ejgallego My point is that I don't get the errors you're seeing, so it must be something specific to the version of cygwin you're using / the packages you install.

Ah, sorry @JasonGross , I was not sure if you had tried this branch; we need help from the platform people then; meanwhile I guess I will make this patch only be used on OSX.

**Comment by ejgallego** [comment]
So the history of this PR is:
- at first it always used `coq_flock`, that worked fine in osx / linux , but failed to build on windows in a strange way
- I tried to alleviate just OSX users [as cygwin flock seems fine], however I am lost with make again

So any help to fix this PR is appreciated.

**Comment by JasonGross** [comment]
I think you should just test to see if flock is available and build it if not.  Something like
```make
FLOCK ?= flock
ifeq (,$(shell which $(FLOCK) 2>/dev/null))
FLOCK := path/to/Coq/flock
Relevant targets: $(FLOCK)
$(FLOCK): flock sources
     recipe to build Coq flock
endif
```
If you want you also do have a USELOCALFLOCK variable or configure argument or something that also enables this target.

File: Makefile.common [hunk 1]
```diff
@@ -27,10 +27,19 @@ CBIN=_build/install/default/bin
...
+FLOCK=_build_vo/default/tools/flock/coq_flock_bin.exe
 
+ifeq ($(ARCH),Darwin)
 $(FLOCK): tools/flock/coq_flock.ml tools/flock/flock.c
 	$(HIDE)dune build $@
+	cp -a $(BCONTEXT)/tools/flock/coq_flock.exe $@
+else
+$(FLOCK):
+	@flock --version >/dev/null 2>&1 || (echo "Error: cannot find flock tool"; exit 1)
+	mkdir -p $(shell dirname $@)
```
Comment (JasonGross) [inline Makefile.common:39]: I don't think you need shell here, I think make has this as a built-in

File: Makefile.common [hunk 2]
```diff
@@ -18,6 +18,29 @@
...
+CONTEXT=_build/install/default
+CBIN=_build/install/default/bin
+CSHARE=_build/install/default/share/coq
+
+# FLOCK
+FLOCK=_build_vo/default/tools/flock/coq_flock_bin.exe
+
+ifeq ($(ARCH),Darwin)
+$(FLOCK): tools/flock/coq_flock.ml tools/flock/flock.c
+	$(HIDE)dune build $@
```
Comment (SkySkimmer) [inline Makefile.common:34]: should have a SHOW

**Comment by ejgallego** [comment]
> It seems that dune now fails with

Yup, copy is not needed as tools in `_build_vo` is just a symlink to tools in the install three IIRC.

File: Makefile.common [hunk 3]
```diff
@@ -18,6 +18,29 @@
...
+
+_DBUILD_DEPS:=
+
+# FLOCK
+ifeq (,$(shell command -v flock 2>/dev/null))
+FLOCK:=_build/default/tools/flock/coq_flock.exe
+_DBUILD_DEPS+=$(FLOCK)
+
+$(FLOCK): tools/flock/coq_flock.ml tools/flock/flock.c
+	$(SHOW)'DUNE    $@'
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply Makefile.common]: Fixed thanks.

**Comment by ejgallego** [comment]
> The CI was failing when I last looked at this, but now it's passing. What changed?

@SkySkimmer fixed it, thanks Gaëtan ; squashed .

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Comment by MSoegtropIMC** [comment]
Should with this merge coq.dev compile again on all platforms, so that I can pick up my work on Coq platform, or are there remaining known issues?

**Comment by ejgallego** [comment]
@MSoegtropIMC , this relieves the need to install the flock binary on osx, but other than that no change here.

What other issues did you have?

**Pull request merged**